src/jaxsim/rbda/contacts/relaxed_rigid.py

```diff
@@ -271,34 +273,33 @@ def compute_contact_forces(
             joint_force_references=joint_force_references,
         )
 
-        def _detect_contact(x: jtp.Array, y: jtp.Array, z: jtp.Array) -> jtp.Array:
-            x, y, z = jax.tree.map(jnp.squeeze, (x, y, z))
-
-            n̂ = self.terrain.normal(x=x, y=y).squeeze()
-            h = jnp.array([0, 0, z - model.terrain.height(x=x, y=y)])
-
-            return jnp.dot(h, n̂)
-
         # Compute the position and linear velocities (mixed representation) of
         # all collidable points belonging to the robot.
         position, velocity = js.contact.collidable_point_kinematics(
             model=model, data=data
         )
 
-        # Compute the activation state of the collidable points
-        δ = jax.vmap(_detect_contact)(*position.T)
+        # Compute the penetration depth and velocity of the collidable points.
+        # Note that this function considers the penetration in the normal direction.
+        δ, δ̇, n̂ = jax.vmap(common.compute_penetration_data, in_axes=(0, 0, None))(
+            position, velocity, model.terrain
+        )
+
+        # Compute the position and velocity in the constraint frame.
+        position_constraint = jax.vmap(lambda δ, n̂: -δ * n̂)(δ, n̂)
+        velocity_constraint = jax.vmap(lambda δ̇, n̂: -δ̇ * n̂)(δ̇, n̂)
```

Comment:
Note that this contact model calls _constraint frame_ what we call $C[W]$, at least to my understanding. cc @flferretti